src/dataflow/src/render/reduce.rs

```diff
@@ -31,23 +76,427 @@ use super::context::Context;
 use crate::render::context::Arrangement;
 use crate::render::datum_vec::DatumVec;
 
-// This enum indicates what class of reduction each aggregate function is.
-#[derive(Clone, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
+/// This enum represents the three potential classes of aggregations.
+#[derive(Copy, Clone, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
 enum ReductionType {
-    // Accumulable functions can be subtracted from (are invertible), and associative.
-    // We can compute these results by moving some data to the diff field under arbitrary
-    // changes to inputs. Examples include sum or count.
+    /// Accumulable functions can be subtracted from (are invertible), and associative.
+    /// We can compute these results by moving some data to the diff field under arbitrary
+    /// changes to inputs. Examples include sum or count.
     Accumulable = 1,
-    // Hierarchical functions are associative, which means we can split up the work of
-    // computing them across subsets. Note that hierarchical reductions should also
-    // reduce the data in some way, as otherwise rendering them hierarchically is not
-    // worth it. Examples include min or max.
+    /// Hierarchical functions are associative, which means we can split up the work of
+    /// computing them across subsets. Note that hierarchical reductions should also
+    /// reduce the data in some way, as otherwise rendering them hierarchically is not
+    /// worth it. Examples include min or max.
     Hierarchical = 2,
-    // Basic, for lack of a better word, are functions that are neither accumulable
-    // nor hierarchical. Examples include jsonb_agg.
+    /// Basic, for lack of a better word, are functions that are neither accumulable
+    /// nor hierarchical. Examples include jsonb_agg.
     Basic = 3,
 }
 
+/// A `ReducePlan` provides a concise description for how we will
+/// execute a given reduce expression.
+///
+/// The provided reduce expression can have no
+/// aggregations, in which case its just a `Distinct` and otherwise
+/// it's composed of a combination of accumulable, hierarchical and
+/// basic aggregations.
+///
+/// We want to try to centralize as much decision making about the
+/// shape / general computation of the rendered dataflow graph
+/// in this plan, and then make actually rendering the graph
+/// be as simple (and compiler verifiable) as possible.
+#[derive(Clone, Debug)]
+enum ReducePlan {
+    /// Plan for not computing any aggregations, just a simple distinct.
```

Comment:
Minor, but could note that it is "just determining the distinct keys", as it's easy to forget that this is also an output of the computation.